arrow/table.go

```diff
@@ -12,49 +12,105 @@ type TableAppender interface {
 	AppendFloats(j int, values []float64) error
 	AppendStrings(j int, values []string) error
 	AppendTimes(j int, values []values.Time) error
+
+	AppendNil(j int) error
+	SetNil(i, j int) error
 }
 
 func AppendBools(tbl TableAppender, j int, vs *array.Boolean) error {
 	var buf [1]bool
+
 	for i := 0; i < vs.Len(); i++ {
-		buf[0] = vs.Value(i)
-		if err := tbl.AppendBools(j, buf[:]); err != nil {
-			return err
+		if vs.IsNull(i) {
+			if err := tbl.AppendNil(j); err != nil {
+				return err
+			}
+		} else {
+			buf[0] = vs.Value(i)
+			if err := tbl.AppendBools(j, buf[:]); err != nil {
+				return err
+			}
 		}
 	}
+
 	return nil
 }
 
 func AppendInts(tbl TableAppender, j int, vs *array.Int64) error {
-	return tbl.AppendInts(j, vs.Int64Values())
+	err := tbl.AppendInts(j, vs.Int64Values())
+	if err != nil {
+		return err
+	}
+
+	return setNils(tbl, j, vs)
 }
 
 func AppendUInts(tbl TableAppender, j int, vs *array.Uint64) error {
-	return tbl.AppendUInts(j, vs.Uint64Values())
+	err := tbl.AppendUInts(j, vs.Uint64Values())
+	if err != nil {
+		return err
+	}
+
+	return setNils(tbl, j, vs)
 }
 
 func AppendFloats(tbl TableAppender, j int, vs *array.Float64) error {
-	return tbl.AppendFloats(j, vs.Float64Values())
+	err := tbl.AppendFloats(j, vs.Float64Values())
+	if err != nil {
+		return err
+	}
+
+	return setNils(tbl, j, vs)
 }
 
 func AppendStrings(tbl TableAppender, j int, vs *array.Binary) error {
 	var buf [1]string
+
 	for i := 0; i < vs.Len(); i++ {
-		buf[0] = vs.ValueString(i)
-		if err := tbl.AppendStrings(j, buf[:]); err != nil {
-			return err
+		if vs.IsNull(i) {
+			if err := tbl.AppendNil(j); err != nil {
+				return err
+			}
+		} else {
+			buf[0] = vs.ValueString(i)
+			if err := tbl.AppendStrings(j, buf[:]); err != nil {
+				return err
+			}
 		}
 	}
+
 	return nil
 }
 
 func AppendTimes(tbl TableAppender, j int, vs *array.Int64) error {
 	var buf [1]values.Time
+
 	for i := 0; i < vs.Len(); i++ {
-		buf[0] = values.Time(vs.Value(i))
-		if err := tbl.AppendTimes(j, buf[:]); err != nil {
-			return err
+		if vs.IsNull(i) {
+			if err := tbl.AppendNil(j); err != nil {
+				return err
+			}
+		} else {
+			buf[0] = values.Time(vs.Value(i))
+			if err := tbl.AppendTimes(j, buf[:]); err != nil {
+				return err
+			}
 		}
 	}
+
+	return nil
+}
+
+func setNils(tbl TableAppender, j int, vs array.Interface) error {
+	if vs.NullN() > 0 {
+		for i := 0; i < vs.Len(); i++ {
+			if vs.IsNull(i) {
+				if err := tbl.SetNil(i, j); err != nil {
```

Comment:
this won't be correct under some conditions, since the column-size may be != `len(vs)` if we just appended data to an existing table.   what we need is for the table-builder to be null-aware for all exported API's.

Reply:
Yeah, in the last commit I pushed a better solution, but it appends elements one by one... Definitely need to put these utilities directly in the builder so that you can get the row offset for that particular column and set nils accordingly.